### ISSUE TITLE
matmul(activation="exp"/"log"/"log2"/"log10"/"log1p") silently selects the approximate kernel while ttnn.exp defaults to accurate — and no string spelling reaches the accurate one

### ISSUE THREAD

**Issue by singhharsh1708** [tenstorrent/tt-metal#55265]
### Summary

`ttnn.matmul(a, b, activation="exp")` silently runs the fast approximate exponential, while `ttnn.exp` on the same tensor runs the accurate one. The same applies to `"log"`, `"log1p"`, `"log2"` and `"log10"`. There is no string spelling that selects the accurate kernel for any of them.

### The table's own convention

In `ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp`, every other entry carrying the approximate-mode flag uses plain name = accurate, with a separate `*_approx` spelling for the fast one:

```
 991:  GELU,    static_cast<float>(false)      993:  if (name == "gelu_approx")
1003:  SIGMOID, ... static_cast<float>(false)  1005:  if (name == "sigmoid_approx")
1083:  MISH,    static_cast<float>(false)      1085:  if (name == "mish_approx")
1015:  SQRT,    static_cast<float>(false)
1018:  RSQRT,   static_cast<float>(false)
1033:  TANH,    static_cast<float>(false)
```

The exp/log family goes the other way, and none of them has an `*_approx` sibling:

```
1021:  EXP,   static_cast<float>(true)
1027:  LOG,   static_cast<float>(true)
1030:  LOG1P, static_cast<float>(true)
1036:  LOG2,  static_cast<float>(true)
1039:  LOG10, static_cast<float>(true)
```

### Both other definitions of the default say false

The LLK template default (`tt_metal/hw/inc/api/compute/eltwise_unary/exp.h:33-37`):

```cpp
template <
    bool approx = false,
    uint32_t scale = 0x3F800000,
    ...
ALWI void exp_tile_init() {
```

and the Python-visible eltwise default (`unary_nanobind.cpp:317`, `unary.hpp:109`):

```cpp
nb::arg("fast_and_approximate_mode") = false,
```

documented as *"Enables a performance-optimized approximation method. When True, the operation runs faster but may produce results with minor precision differences. Defaults to `False`."*

So the string-activation path opts into a mode the op's own API documents as off by default, without saying so.

### What the approximate path is

`tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_exp.h:444-461` — a Schraudolph-style bit trick with 3 fractional bits, not a polynomial:

```cpp
if constexpr (APPROXIMATION_MODE) {
    constexpr int FRAC_BITS = 3;
    constexpr std::uint32_t SP_BIAS = 127 << FRAC_BITS;
    sfpi::vFloat conv = in * vConstLn2Recip;      // * 1/ln2
    sfpi::vInt tmp = sfpi::as<sfpi::vInt>(conv) - c23_73;
    tmp += SP_BIAS;
    out = sfpi::as<sfpi::vFloat>(tmp << (10 - FRAC_BITS));   // shift into exponent
} else {
    out = _sfpu_exp_(sfpi::setsgn(in, 0));
    v_if(in < 0) { out = sfpu_reciprocal_iter<2>(out); }
    v_endif;
}
```

With 3 fractional bits the error is a percent-scale relative error, periodic in `x/ln2` — orders of magnitude beyond bfloat16 rounding, and it does not average out across a tensor. Measured on **ttsim** (tenstorrent/ttsim v1.10.3, `libttsim_bh.so`, Blackhole) with the prebuilt tt-metal `v0.79.0-dev20260902` ubuntu22.04 debs and SFPI 7.72.0. No hardware is required to reproduce this:

```
a = ones([1,1,32,32]) bf16, b = eye(32) bf16   (so the product is 1.0)
ttnn.exp(ttnn.matmul(a, b))           -> 2.71875      # bf16(e), accurate path
ttnn.matmul(a, b, activation="exp")   -> 2.796875     # approximate path
```

The two spellings return different values for the same input, confirming that the string activation selects the approximate kernel while `ttnn.exp` selects the accurate one.

### Reach

Same divergence through `ttnn.linear`, and through `ttnn.experimental.dit_rms_norm_unary_fused`, whose binding docstring (`dit_rms_norm_unary_fused_nanobind.cpp:53-54`) and header comment (`dit_rms_norm_unary_fused.hpp:16-18`) both state it is *"Equivalent to `ttnn.<activation>(ttnn.rms_norm(x, ...))`"*. For `activation="exp"` that written equivalence does not hold.

For `log`, the approximate path additionally skips normalisation, so `-0.0` and subnormal inputs do not return `-inf` / the correct value.

No test covers these strings. `tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py` rewrites its string cases into parameterless `UnaryWithParam` objects before the call, so the table is never exercised on device.

### Fix

Flip the five constants so the plain names match the LLK default and the rest of the table:

```cpp
:1021  UnaryWithParam(UnaryOpType::EXP,   static_cast<float>(false))
:1027  UnaryWithParam(UnaryOpType::LOG,   static_cast<float>(false))
:1030  UnaryWithParam(UnaryOpType::LOG1P, static_cast<float>(false))
:1036  UnaryWithParam(UnaryOpType::LOG2,  static_cast<float>(false))
:1039  UnaryWithParam(UnaryOpType::LOG10, static_cast<float>(false))
```

and optionally add `exp_approx` / `log_approx` / `log2_approx` / `log10_approx` / `log1p_approx` returning `true`, matching the existing `gelu_approx` / `sigmoid_approx` / `mish_approx` pattern so callers who want the fast kernel keep it. No kernel change, no new dispatch.

If the current behaviour is deliberate for performance, then the docstring equivalence claim on the fused ops should be corrected instead, and an accurate spelling added.

All file:line against `2aa657a`.

I would like to work on this through the bounty program, and am happy to prepare the PR.

**Comment by singhharsh1708** [comment]
Measured on **ttsim** (tenstorrent/ttsim v1.10.3, `libttsim_bh.so`, Blackhole) with the prebuilt tt-metal `v0.79.0-dev20260902` ubuntu22.04 debs and SFPI 7.72.0 — no hardware.

```
a = ones([1,1,32,32]) bf16, b = eye(32) bf16   (so the product is 1.0)
ttnn.exp(ttnn.matmul(a, b))           -> 2.71875      # bf16(e), accurate path
ttnn.matmul(a, b, activation="exp")   -> 2.796875     # approximate path
```

The two spellings return different values for the same input, confirming that the string activation selects the approximate kernel while `ttnn.exp` selects the accurate one.

**Comment by MusounoEnma** [comment]
Hi @jberkowitzTT and maintainers, I would like to request assignment to this issue under the bounty program.

### Root Cause & Implementation Plan:
1. **Align Default Activation with Accurate Mode (`unary_op_utils.cpp`)**:
   - Update `EXP`, `LOG`, `LOG1P`, `LOG2`, and `LOG10` entries in `string_to_unary_with_param` to use `static_cast<float>(false)` by default, matching `ttnn.exp`'s accurate behavior and the convention used by `GELU`, `SIGMOID`, and `MISH`.
   - Add `exp_approx`, `log_approx`, `log1p_approx`, `log2_approx`, and `log10_approx` activation strings mapping to `static_cast<float>(true)` so existing callers requiring the fast approximate kernel retain that option.

2. **Testing (`tests/ttnn/unit_tests/operations/matmul/test_matmul_activation.py`)**:
   - Add unit tests verifying accurate output parity against `torch.matmul` followed by `torch.exp` / `torch.log`.
   - Verify that passing `exp_approx` / `log_approx` preserves fast approximate execution.

Ready to implement and submit the PR upon assignment!

**Comment by singhharsh1708** [comment]
@MusounoEnma thanks, but I'm the reporter here and the issue body already states I'd like to take this through the bounty program and prepare the PR — the plan you've outlined is the fix proposed in the report above.

@jberkowitzTT I have a working reproduction for this (ttsim v1.10.3 on the published debs, no hardware), which is how the measured `2.71875` vs `2.796875` numbers in the body were produced, so I can validate the change end to end and add the regression coverage.
